core/services/gateway/connector/connector.go

```diff
@@ -0,0 +1,206 @@
+		}
+		gateways[gw.Id] = gateway
+	}
+	connector.gateways = gateways
+	return connector, nil
+}
+
+func (c *gatewayConnector) SendToGateway(ctx context.Context, gatewayId string, msg *gateway.Message) error {
+	data, err := c.codec.EncodeResponse(msg)
+	if err != nil {
+		return fmt.Errorf("error encoding response for gateway %s: %v", gatewayId, err)
+	}
+	gateway, ok := c.gateways[gatewayId]
+	if !ok {
+		return fmt.Errorf("invalid Gateway ID %s", gatewayId)
+	}
+	if gateway.conn == nil {
+		return fmt.Errorf("connector not started")
+	}
+	return gateway.conn.Write(ctx, websocket.BinaryMessage, data)
+}
+
+func (c *gatewayConnector) readLoop(gatewayState *gatewayState) {
+	for {
+		select {
+		case <-c.shutdownCh:
+			c.closeWait.Done()
+			return
+		case item := <-gatewayState.conn.ReadChannel():
+			msg, err := c.codec.DecodeRequest(item.Data)
+			if err != nil {
+				c.lggr.Error("parse error when reading from Gateway ", gatewayState.config.Id, err)
+				break
+			}
+			c.handler.HandleGatewayMessage(gatewayState.config.Id, msg)
+		}
+	}
+}
+
+func (c *gatewayConnector) reconnectLoop(gatewayState *gatewayState) {
+	for {
+		conn, err := gatewayState.wsClient.Connect(context.Background(), gatewayState.url)
+		if err != nil {
+			c.lggr.Error("connection error!")
+		} else {
+			closeCh := gatewayState.conn.Restart(conn)
+			<-closeCh
+			c.lggr.Info("connection closed")
+		}
+		select {
+		case <-c.shutdownCh:
+			c.closeWait.Done()
+			return
+		case <-time.After(time.Duration(c.config.ReconnectDelayMillis) * time.Millisecond):
+			c.lggr.Info("reconnecting ...")
+		}
+	}
+}
+
+func (c *gatewayConnector) Start(ctx context.Context) error {
+	return c.StartOnce("GatewayConnector", func() error {
+		c.lggr.Info("starting gateway connector")
+		for _, gatewayState := range c.gateways {
+			gatewayState.conn = network.NewWSConnectionWrapper()
+			go c.readLoop(gatewayState)
+			go c.reconnectLoop(gatewayState)
+		}
+		c.closeWait.Add(2 * len(c.gateways))
+		return nil
+	})
+}
+
+func (c *gatewayConnector) Close() error {
+	return c.StopOnce("GatewayConnector", func() (err error) {
+		c.lggr.Info("closing gateway connector")
+		close(c.shutdownCh)
+		for _, gatewayState := range c.gateways {
+			gatewayState.conn.Close()
+		}
+		c.closeWait.Wait()
+		return nil
+	})
+}
+
+func (c *gatewayConnector) NewAuthHeader(url *url.URL) ([]byte, error) {
+	authHeaderElems := &network.AuthHeaderElems{
+		Timestamp:  uint32(c.clock.Now().Unix()),
+		DonId:      c.config.DonId,
+		GatewayURL: url.String(),
+	}
+	packedElems := network.Pack(authHeaderElems)
+	signature, err := c.signer.Sign(packedElems)
+	if err != nil {
+		return nil, err
+	}
+	return append(packedElems, signature...), nil
+}
+
+func (c *gatewayConnector) ChallengeResponse(challenge []byte) ([]byte, error) {
+	if len(challenge) < c.config.MinHandshakeChallengeLen {
+		return nil, errors.New("handshake challenge too short")
```

Comment:
For most errors, I suggest to create exported vars. You may not see a value of doing this until you start adding negative unit tests. However, higher-order components (or their tests) may need to check against specific errors.

Reply:
Will do separately. Errors might change if anything changes in the handshake.